dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoaderRef.java

```diff
@@ -0,0 +1,160 @@
+/*******************************************************************************
+ * Copyright (c) 2025 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.classloading.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.ref.WeakReference;
+import java.net.URL;
+import java.security.ProtectionDomain;
+import java.util.EnumSet;
+import java.util.Enumeration;
+import java.util.Objects;
+
+import org.osgi.framework.Bundle;
+
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.wsspi.classloading.ApiType;
+
+/**
+ * This class is used to create either weak references to AppClassLoaders
+ */
+@Trivial
+public class ClassLoaderRef extends LibertyLoader implements SpringLoader {
+    static {
+        ClassLoader.registerAsParallelCapable();
+    }
+
+    private final WeakReference<AppClassLoader> classLoaderWeakRef;
+
+    ClassLoaderRef(AppClassLoader appClassLoader) {
+        super(appClassLoader.getParent());
+        classLoaderWeakRef = new WeakReference<>(appClassLoader);
+    }
+
+    @Override
+    protected final Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        return loadClass(name, resolve, false, false);
+    }
+
+    @Override
+    public URL getResource(String resName) {
+        AppClassLoader appCL = classLoaderWeakRef.get();
+        return appCL != null ? appCL.getResource(resName) : parent.getResource(resName);
```

Comment:
Overall, I question trying to do anything when `appCL` is `null`.

Reply:
The thread in question will have had the TCCL set to the application's TCCL.  But now that application's class loader is gone.  Isn't that now in a defunct state for the thread?  Before we would have happily tried to load additional application classes, but now that will fail regardless of what we do.

Perhaps we should have a private method `getAppCL` which will post an error (only once) when the return value is `null`.  This way it is obvious in the logs that we have a leaked TCCL and a defunct thread from an application.  I would be bonus points if that error message also gave the application name for the leaked TCCL.

Reply:
Returning null in Classloader methods that execute after the weak appCL reference has been GC'd is my only concern with this PR.  After an application stops it's very possible something other than the app's TCCL holds a strong reference (or chain) to the stale app CL -- e.g. a managed thread local that references a stale webApp instance, which holds a strong reference to its app CL.  I agree w/ Tom's suggestion to log information to identity any stale app CL in use, and the application name is essential for service.  The application name info is in the AppClassLoader's config object (OMP - I'll provide examples from MAT).  On a related note, we need a product issue to set state in the AppCL that enables MAT, trace, introspections, etc. to identify stale app CLs.

Reply:
If there is a strong reference still held somewhere, the WeakReference in ClassLoaderRef will not release, so that isn't a problem.  I added a scenario in the unit test to show the importance of calling the parent because otherwise using ThreadContextClassLoader after the application classloader is garbage collected won't even let you load a Java class if we don't call the parent that was not garbage collected.  I expect the same for resources when trying to load exposed resources by the Liberty runtime for things like ServiceLoader to ind our implementations, we need that to continue to work otherwise it would be a breaking change.